File: desafio2/scala/src/main/scala/Main.scala
import scala.io.Source
import scala.util._
import scala.xml._

object Main {

  val time_report_message = "tiempo ocupado para generar el reporte: "
 // val report_format = "%-30.30s max:%5.1f  min:%5.1f   actual: %5.1f %s"
  val no_cities_provided_message = "debe ingresar una lista de ciudades"

  val api_key = sys.env("WEATHER_API_KEY")
  val maxTries = 10
  val threadSleep = 100

  sealed trait Args
  case object NoArgs extends Args
  case class SeqArgs(args:List[String]) extends Args
  case class ParArgs(args:List[String]) extends Args

  def checkArgs(args: Array[String]) : Args = 
    if (args.size == 0) 
      NoArgs
    else if (args(0) != "-p")
        SeqArgs(args.toList)
    else if (args.toList.tail.isEmpty)
      NoArgs
    else
      ParArgs(args.toList.tail)

  def makeUrl(city:String) : String = s"http://api.openweathermap.org/data/2.5/weather?q=$city&mode=xml&units=metric&appid=$api_key&lang=sp"

  sealed trait WeatherResult
  case class Error(error: String, city: String) extends WeatherResult
  case class WeatherReport(city: String, temp: Double, min: Double, max: Double, conditions: String) extends WeatherResult

  def parseDouble(str: String) : Double = Try(str.toDouble).getOrElse(0.0)

  def parseApiResponse(city: String, xmlResponse: String): WeatherResult = {
    val xml = XML.loadString(xmlResponse)
    val current = xml \\ "current"
    if (current.isEmpty)
      Error("error parsing xml response", city)
    else {
      val city = xml \\ "city" \ "@name"
      val temp = xml \\ "temperature" \ "@value"
      val min  = xml \\ "temperature" \ "@min"
      val max  = xml \\ "temperature" \ "@max"
      val weather = xml \\ "weather" \ "@value"
      WeatherReport(city.text, parseDouble(temp.text), parseDouble(min.text), parseDouble(max.text), weather.text)
    }
  }

  def apiCall(city: String): WeatherResult = {
    val url = makeUrl(city)
    for (tries <- 1 to maxTries) {
      val response = Try(Source.fromURL(url))
      response match {
        case Success(rsp) =>
          return parseApiResponse(city, rsp.mkString)
        case Failure(e) =>
          Thread.sleep(threadSleep)
      }
    }
    Error("error descargando url", city)
  }

  def compareWeatherReports(rep1: WeatherResult, rep2: WeatherResult) : Boolean =
    (rep1, rep2) match {
      case (_, Error(_,_)) => false
      case  (Error(_,_),_) => false
      case (WeatherReport(_, temp1, _, _, _), WeatherReport(_, temp2, _, _, _))  => temp2 < temp1
    } 

  def printReports(reports:List[WeatherResult]) : Unit = 
    reports.sortWith(compareWeatherReports).foreach { 
      case Error(err, city) =>
        println(s"$city Error: $err")
      case WeatherReport(city, temp, min, max, conditions) =>
        println(f"$city%-30.30s max:$max%5.1f  min:$min%5.1f   actual: $temp%5.1f $conditions")
    }


  def seqFetch(cities: List[String]) : Unit = 
    printReports(cities.map(city => apiCall(city)))

  def parFetch(cities:List[String]) : Unit = {
    val reports = (cities.par map (city => apiCall(city))).toList
    printReports(reports)
  }

  def showTime(block : => Unit) : Unit = {
    val t0 = System.nanoTime()
    val result = block
    val t1 = System.nanoTime()
    val elapsed = t1 - t0
    val msecs = elapsed / 1000000
    val hours = msecs / 3600000
    val mins = (msecs % 3600000) / 60000
    val secs = ((msecs % 3600000) % 60000) / 1000.0

    println(f"$time_report_message $hours%d:$mins%02d:$secs%02.3f")
  }

  def main(args: Array[String]): Unit = 
    checkArgs(args) match {
      case NoArgs => println(no_cities_provided_message)
      case SeqArgs(argList) => showTime { seqFetch(argList) }
      case ParArgs(argList) => showTime { parFetch(argList) }
    } 

}

File: desafio2/rust/src/main.rs
extern crate getopts;
extern crate treexml;
use getopts::Options;
use std::thread;
use std::sync::mpsc;
use std::env;
use treexml::Document;
use std::cmp::Ordering;
use cabot::{RequestBuilder, Client};

enum ApiResult {
    Error{city: String, error: String},
    WeatherReport{city: String, temp: f32, max:f32, min:f32, conditions: String}
}

fn api_call(city:&String, api_key:&String) -> ApiResult {
    let url = format!("http://api.openweathermap.org/data/2.5/weather?q={}&mode=xml&units=metric&appid={}&lang=sp", city, api_key);
    let request = RequestBuilder::new(&url).build().unwrap();
    let client = Client::new();
    for _ in 1..10 {
        match client.execute(&request) { 
            Err(_) => {
                std::thread::sleep(std::time::Duration::from_millis(100)) 

            },
            Ok(res) => {
                let body = res.body_as_string().unwrap();
                match Document::parse(body.as_bytes()) {
                    Err(_) => {
                        std::thread::sleep(std::time::Duration::from_millis(100)) 
                    },
                    Ok(doc) => {
                        let root = doc.root.unwrap();
                        let temp = root.find_child(|tag| tag.name == "temperature").unwrap().clone();
                        let weather = root.find_child(|tag| tag.name == "weather").unwrap().clone();

                        let min_temp: f32= temp.attributes["min"].parse().unwrap();
                        let max_temp: f32= temp.attributes["max"].parse().unwrap();
                        let cur_temp: f32= temp.attributes["value"].parse().unwrap();

                        return ApiResult::WeatherReport{city: city.clone(), 
                            temp: cur_temp, max: max_temp, min:min_temp, 
                            conditions: weather.attributes["value"].clone() }
                    }
                }
            }
        }
    }
    ApiResult::Error{city: city.clone(), error: "error invocando api".to_string()}
}

fn par_fetch(cities:&Vec<String>, api_key:String) -> Vec<ApiResult> {
    let (tx, rx) = mpsc::channel();

    for city in cities {
        let tx = tx.clone();
        let city = city.clone();
        let api_key = api_key.clone();
        thread::spawn(move || {
            let report = api_call(&city, &api_key);
            tx.send(report).unwrap();
        });
    }
    let mut reports : Vec<ApiResult> = Vec::with_capacity(cities.len());
    for _ in cities {
        let rep = rx.recv().unwrap();
        reports.push(rep);
    }
   reports 
}

fn seq_fetch(cities:&Vec<String>, api_key:String) -> Vec<ApiResult> {
    let mut reports : Vec<ApiResult> = Vec::with_capacity(cities.len());
    for city in cities {
        reports.push(api_call(city, &api_key));
    }
    reports 
}

fn comp_api_result(a:&ApiResult, b:&ApiResult) -> Ordering {
    match a {
        &ApiResult::Error{city:_, error:_} => Ordering::Less,
        &ApiResult::WeatherReport{city:_, temp:ta, max:_, min:_, conditions:_} => 
            match b {
                &ApiResult::Error{city:_, error:_} => Ordering::Less,
                &ApiResult::WeatherReport{city:_, temp:tb, max:_, min:_, conditions:_} => 
                    if ta > tb { Ordering::Less } else { Ordering::Greater }
            }
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let mut opts = Options::new();
    opts.optflag("p", "par", "ejecuta la busqueda en paralelo");
    let matches = match opts.parse(&args[1..]) {
    	Ok(m) => { m }
    	Err(f) => { panic!(f.to_string()) }
    };
    let par = matches.opt_present("p");
    let cities = matches.free;
    if cities.is_empty() {
    	println!("debe ingresar una lista de ciudades");
    	return;
    }
    let api_key = env::var("WEATHER_API_KEY").expect("debe definir WEATHER_API_KEY");

    let t0 = std::time::Instant::now();
    let mut result = if par { par_fetch(&cities, api_key) } else { seq_fetch(&cities, api_key) };

    result.sort_by(|a,b| comp_api_result(&a,&b));
    for report in result {
        match report {
            ApiResult::Error{city:c, error:e} => { println!("{:?} Error: {:?}", e, c) }
            ApiResult::WeatherReport{city, temp, max, min, conditions} =>  { println!("{:<30} max:{:5.1}  min:{:5.1}   actual: {:5.1} {}", city, temp, max, min, conditions) }
        }
    }
    let dur = t0.elapsed();
    let secs = dur.as_secs();
    let nanos = dur.subsec_nanos();
    let hours = secs / 3600;
    let mins  = (secs % 3600) / 60;
    let sec  = (secs % 3600) % 60;
    let frac  = nanos/1000000;
    println!("tiempo ocupado para generar el reporte: {} = {}:{}:{}.{}", secs, hours, mins, sec, frac);
    
}
